core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,248 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import io.grpc.Attributes;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.ExperimentalApi;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer2;
+import io.grpc.LoadBalancer2.PickResult;
+import io.grpc.LoadBalancer2.Subchannel;
+import io.grpc.LoadBalancer2.SubchannelPicker;
+import io.grpc.Metadata;
+import io.grpc.NameResolver;
+import io.grpc.ResolvedServerInfo;
+import io.grpc.ResolvedServerInfoGroup;
+import io.grpc.Status;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link LoadBalancer} that provides round-robin load balancing mechanism over the
+ * addresses from the {@link NameResolver}.  The sub-lists received from the name resolver
+ * are considered to be an {@link EquivalentAddressGroup} and each of these sub-lists is
+ * what is then balanced across.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1771")
+public class RoundRobinLoadBalancerFactory2 extends LoadBalancer2.Factory {
+  private static final RoundRobinLoadBalancerFactory2 instance =
+      new RoundRobinLoadBalancerFactory2();
+
+  private RoundRobinLoadBalancerFactory2() {
+  }
+
+  public static RoundRobinLoadBalancerFactory2 getInstance() {
+    return instance;
+  }
+
+  @Override
+  public LoadBalancer2 newLoadBalancer(LoadBalancer2.Helper helper) {
+    return new RoundRobinLoadBalancer(helper);
+  }
+
+  @VisibleForTesting
+  static class RoundRobinLoadBalancer extends LoadBalancer2 {
+    private final Helper helper;
+    private final BiMap<EquivalentAddressGroup, Subchannel> subchannels = HashBiMap.create();
+
+    @VisibleForTesting
+    static final Attributes.Key<AtomicReference<ConnectivityStateInfo>> STATE_INFO =
+        Attributes.Key.of("state-info");
+
+    public RoundRobinLoadBalancer(Helper helper) {
+      this.helper = helper;
+    }
+
+    @Override
+    public void handleResolvedAddresses(List<ResolvedServerInfoGroup> servers,
+        Attributes attributes) {
+      Set<EquivalentAddressGroup> currentAddrs = ImmutableSet.copyOf(subchannels.keySet());
```

Comment:
The `BiMap` will rise `java.util.ConcurrentModificationException` because we're removing from the map over which we're iterating indirectly, few lines below (`subchannel.remove`).

I'll add comment.